Create and update intro-js.md

redixhumayun opened the pull request:
Need to add bootstrap to the file above. Everything else should be up to date.

redixhumayun committed:
Create and update intro-js.md

Need to add bootstrap to the file above. Everything else should be up to date.

redixhumayun committed:
Update to fix mistake

Changed closure to currying

redixhumayun committed:
Update to list items

redixhumayun committed:
Update to fix list items

redixhumayun committed:
Update to code formatting

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+
+## Why is JavaScript weird? 
+
+1. Lack of types
```
Not a fan of this formatting. Try some kind of title like `### Lack of Types`

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+
+1. Lack of types
+
+JavaScript is a *loosely typed* or *dynamically typed* language
```
Explain this

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+var a = 'random' //a is a string here
+var a = true //a is a boolean here
+```
+
````
Add something like `JavaScript determines types during "runtime", in other words, the type is determined when the code is running`

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+
+A solution to this has been provided through [TypeScript](https://www.typescriptlang.org/docs/home.html)
+
+Angular 2 utilizes TypeScript extensively.
```
Change to 4

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+
+This makes JavaScript development cumbersome and bug ridden, but there are improvements being made. 
+
+A solution to this has been provided through [TypeScript](https://www.typescriptlang.org/docs/home.html)
```
`TypeScript is a "strongly typed" language unlike JavaScript`

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+Angular 2 utilizes TypeScript extensively.
+
+2. Lack of classes
+JavaScript did not have the concept of classes when it was first written.
```
`Originally they were implemented by creating multiple functions`

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+2. Lack of classes
+JavaScript did not have the concept of classes when it was first written.
+
+Classes have now been provided as part of [ECMAScript 2015](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes), but this is just syntactic sugar. 
```
Technically they're only syntactic sugar if compiled back to ES5. Also, be sure to explain what syntactic sugar is.

maneesht replied:
Still not addressed.

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+Example class
+```javascript 
+class Shape {
+    constructor(param1, param2) {
````
Explain constructors, how the parameters are technically optional

claytn replied:
Why are we explaining classes before explaining functions and objects?

redixhumayun replied:
Not entirely sure, but maybe because we're assuming a technical background  of some programming language for the students? Would we need to explain how functions and objects work in that case?

claytn replied:
I'm just thinking in terms of learning javascript from the ground up. It is pretty rare that people use javascript with an object oriented approach when using it for web development. The only time I ever use classes is with React and even with React I try to use functional components whenever I can. I think teaching some es6 concepts are important, but I think it's more important to know vanilla javascript before teaching what is literally pseudo classes. @maneesht what are your thoughts on this?

maneesht replied:
I think it's important teaching new students the latest version of JavaScript. No one uses ES5 anymore. Learning a new language I'd like to learn classes first, then objects, but I'd like to know functions before classes.
Functions -> Classes -> Objects

claytn replied:
That works

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+
+This one is a little hard to wrap your heads around so we won't go into too much detail, but it is one of the most fundamental aspects of JavaScript. It is what makes JavaScript such a powerful language. 
+
+Basically, what this means is that functions can be constructed during execution, they can be passed around as variables or parameters to other functions and you can also return a function as the value of another function (what is referred to as a currying). 
```
Meat is here, and since this is a very difficult topic to cover, I'm not sure if it's 100% necessary

redixhumayun replied:
Okay, do you want me to go ahead and remove it?

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+## Basic DataTypes in JavaScript
+Some of the ECMAScript standard datatypes:
+* Boolean 
+* Null
```
Datatype?

redixhumayun replied:
I'll just go with data types

maneesht replied:
Technically `Undefined` and `null` are data types, but I see them more as values and mentioning them along with `Boolean` and `String` may confuse people

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+Some of the ECMAScript standard datatypes:
+* Boolean 
+* Null
+* Undefined
```
^

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+To get the name from the above object, you would simply do
+
+```javascript
+    var name = jsonObj.name //Zaid Humayun
````
Let's consistently use semicolons.

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+Setting some value up as a boolean is fairly simple
+
+```javascript
+var a = true //a is the true boolean value
````
semicolon

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+
+```javascript
+var a = true //a is the true boolean value
+var b = false //b is the false boolean value
````
here as well

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+var myStr = '' //defining an empty string here
+if(myStr) {
+    //this branch will not fire
+}else {
```
`} else {` space between closing `}` and `else`

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+    //this branch will fire
+}
+
+var a = 2 //defining a as a number here
```
semicolon

maneesht commented on intro-js.md:
```diff
@@ -0,0 +1,149 @@
+var a = 2 //defining a as a number here
+if(a) {
+    //this branch will fire
+}else {
```
formatting

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+It is possible to typecast an object (for example) to a boolean type using a double not operator (that is, !!)
+
+```javascript 
+oObject = {'key': value} //some object here
````
Did you mean `oObject`?

redixhumayun replied:
For the variable name? Yeah, I meant that. Want me to change the variable name?

maneesht replied:
Yes, `oObject` seems like an unfortunate choice for a variable name

maneesht commented on intro-js.md:
````diff
@@ -0,0 +1,149 @@
+
+```javascript 
+oObject = {'key': value} //some object here
+val = !!oObject; //typecasting to boolean to set another value
````
I'm debating whether we should add this or not...

redixhumayun replied:
Yeah, I was debating that as well. It seems like a complex topic, and I'm sure most people will just ignore it and move on. But, I'd suggest leaving it in there just for reference.

maneesht replied:
Let's leave it for now. We'll see how it fits later.

maneesht reviewed (changes requested):
Take a look at what I've commented. 
Do we plan on including functions this time or next meeting?

maneesht committed:
Modified bootstrap css

redixhumayun committed:
Implemented suggested changes

Left in functions as first class objects section and the double not operator. Still need to make a call about that.

maneesht commented on intro-js.md:
```diff
@@ -34,17 +47,40 @@ class Shape {
 
-This one is a little hard to wrap your heads around so we won't go into too much detail, but it is one of the most fundamental aspects of JavaScript. It is what makes JavaScript such a powerful language. 
+#### Side Note
+What exactly is syntactic sugar? 
```
Mentioned too late. I'd add this a as a note right when syntactic sugar is mentioned above.

redixhumayun replied:
I've been thinking about the compiled back to the ES5 bit. Do we want to include that? Will they understand what compiling back to ES5 means?

maneesht reviewed (changes requested):
Close, but not quite there yet. Good work so far!

redixhumayun committed:
Update bootstrap-css.md

Included link to bootstrap docs, created section for offset of columns, and added a resources section.

redixhumayun committed:
Update intro-js.md

Moved location of syntactic sugar side note, removed undefined and null from data types list, changed oObject to obj.

maneesht commented on intro-js.md:
```diff
@@ -38,6 +38,13 @@ Originally they were implemented by creating multiple functions.
+#### Side Note
+What exactly is syntactic sugar? 
+
+Syntactic sugar is basically syntax for a programming language that makes it easier for developers to read and write. 
```
We need something a little bit more in-depth than this. 
What I was thinking was more along the lines of:
First introduce ECMAScript, the different versions, then explain how compatibility of each version is dependent on the browser. 
After that, we can mention that syntactic sugar is `helpful syntax to make developer's lives easier for development, but in the end, in terms of functionality differences, there is none`

redixhumayun replied:
Okay, I've added more to it but I'm not sure if its just going to confuse them more or not. I feel like we might be throwing too much at them with the ECMAScript standards, syntactic sugar and tipping them off to how classes are actually defined in JavaScript. That's a whole lesson in and of itself. 
Maybe it might be best to just show them how to define a class in ES6 and leave it at that.

maneesht commented on bootstrap-css.md:
```diff
@@ -0,0 +1,82 @@
+* Stylized Buttons
+* Responsive Navbar
+* Responsive Scaffolding System
+
```
We should probably add a section here saying how to install and set up bootstrap. I was thinking they just use the CDN for now. Thoughts?

redixhumayun replied:
Yeah, just the CDN should be good. You'll have to introduce npm to them before teaching them how to install it locally.

redixhumayun committed:
Update bootstrap-css.md

Added section on how to set up bootstrap in html file

redixhumayun committed:
Update intro-js.md

Added additional information for the side note

maneesht commented:
Take a look at slides.com/maneeshtewani/deck-10

redixhumayun commented:
Slides look good. I noticed that we haven't shown them how to link a HTML page to a JavaScript file. Should we include that?

claytn commented on intro-js.md:
```diff
@@ -0,0 +1,222 @@
+
+But there's something weird about booleans in JS too. Who could have guessed right? 
+
+JavaScript has something called *truthy* and *falsey* values. 
```
I wouldn't have thought about adding this, but I think truthy and falsey values are great to bring up earlier in teaching javascript like this. Nice thinking!

claytn commented on intro-js.md:
```diff
@@ -0,0 +1,222 @@
+* Number
+* String
+
+# JSON
```
What are your thoughts on explaining regular js objects first and then bringing up json?

redixhumayun replied:
Sounds like a good idea. How in depth were you thinking?

claytn replied:
Nothing too crazy. I think object literals can be taught easily, but let's not get into using the new keyword or anything to do with the prototype chain for the time being. I agree that the oop side of javascript is valuable to know, but for making web apps it's rare that those concepts are ever needed.

redixhumayun replied:
This sounds good to me. @maneesht What do you think?

maneesht replied:
I think that's okay. Let's get some of these ideas written up so we can see what it looks like...this is ending up as a megathread

redixhumayun replied:
Okay, done with this. Take a look and let me know if its enough.

redixhumayun committed:
Update intro-js.md

Added info about javascript objects. Nothing using the new keyword or the prototype chain.

maneesht committed:
Small clarity changes

maneesht committed:
Fixed bootstrap issues

maneesht committed:
Cleanup of file structure

maneesht committed:
Included folders

maneesht commented:
Merged.
